src/main/scala/is/hail/methods/LinearMixedRegression.scala

```diff
@@ -163,70 +187,139 @@ object LinearMixedRegression {
   }
 }
 
+
 object DiagLMM {
   def apply(
     C: DenseMatrix[Double],
     y: DenseVector[Double],
     S: DenseVector[Double],
+    U: DenseMatrix[Double],
     optDelta: Option[Double] = None,
     useML: Boolean = false): DiagLMM = {
 
-    require(C.rows == y.length)
+    new DiagLMMSolver(C, y, S, U, optDelta, useML).solve()
+  }
+}
 
-    val (delta, optGlobalFit) = optDelta match {
-        case Some(d) => (d, None)
-        case None =>
-          val (d, gf) = fitDelta(C, y, S, useML)
-          (d, Some(gf))
-      }
 
-    val n = y.length
+class DiagLMMSolver(
+  C: DenseMatrix[Double],
+  y: DenseVector[Double],
+  S: DenseVector[Double],
+  U: DenseMatrix[Double],
+  optDelta: Option[Double] = None,
+  useML: Boolean = false) {
+
+  val Ut = U.t
+  val UtC = Ut * C
+  val Uty = Ut * y
+
+  require(UtC.rows == Uty.length)
+
+  val CtC = C.t * C
+  val Cty = C.t * y
+  val yty = y.t * y
+
+  val n = y.length
+  val d = C.cols
+
+  val (delta, optGlobalFit) = optDelta match {
+    case Some(del) => (del, None)
+    case None =>
+      val (del, gf) = fitDelta()
+      (del, Some(gf))
+  }
+
+  def solve(): DiagLMM = {
+    val invD = (S + delta).map(1 / _)
+    val dy = Uty :* invD
+    var ydy = Uty dot dy
+    var Cdy = UtC.t * dy
+    var CdC = UtC.t * (UtC(::, *) :* invD)
+
+    if(S.length < n) {
+      ydy = ydy + (1.0 / delta) * (yty - Uty.t * Uty)
```

Comment:
I fixed all the occurrences where I was recomputing the thing independent of delta. Doing vector divided by scaler is easy. It's just ``vec / delta``. Elementwise is for when you have two vectors and you want to divide corresponding elements by each other.